Always Look On The Bright Side of the Annex

ariard opened the pull request:
Build on top  of #4.

This PR proposes a validation format for the taproot [annex](https://github.com/ariard/bips/blob/2022-07-bip-annex/bip-annex.mediawiki). It allows to extend the usual transaction fields with new data records allowing witness signatures to commit to them. The data records can be subject to new validation rules.

So what's the PR standard we're looking for new softfork idea? There is a WIP BIP available here:
https://github.com/ariard/bips/blob/2022-07-bip-annex/bip-annex.mediawiki

List of following use-cases that could be supported with the annex:
- per-input lock-time
- historical block hash commitment
- sighash_group
- pay-with-neighboor-txid
- block-feerate lock-point
- "tx size is smaller than K" 
- more?

ajtowns commented:
Is there any reason not to PR the draft BIP to the bips repo? Non "Final" BIPs can be updated essentially at will, as I understand it...

I don't think you can have a package weight be consensus enforced (if you just see the block, how do you know what the package was); so that feature would be more "package weight committed to by signature", I think.

ariard commented:
> Is there any reason not to PR the draft BIP to the bips repo? Non "Final" BIPs can be updated essentially at will, as I understand it...

Yeah I need to refine some BIP sections though then no reason not to PR the draft to the bips repo.

> I don't think you can have a package weight be consensus enforced (if you just see the block, how do you know what the package was); so that feature would be more "package weight committed to by signature", I think.

Maybe you can introduce the notion of package at the consensus-level, you validate the whole chain of transaction since the latest UTXO set. Though that sounds gross. Yeah "package weight committed to by signature" could work as would be a restraint enforced from the spent scriptpubkey.

instagibbs commented:
So to my mind there's two parts:
1) the format, which can be opted into prior to any softfork
2) the softfork, which would *require* interpretation of the annex in a certain way, with or without consensus meaning for the actual values

is this correct?

ariard commented:
>  the format, which can be opted into prior to any softfork
>    the softfork, which would require interpretation of the annex in a certain way, with or without consensus meaning for the actual values

Yes, once we define a format at the policy-level, if we would like to enforce another data format at the consensus-level, all the annex signers would need to upgrade their data fields, as otherwise their transactions could be rejected for being invalid.

ajtowns commented:
> So what's the PR standard we're looking for new softfork idea?

CI passing at a minimum :)

```
script/interpreter.cpp:2117:69: error: narrowing conversion of ‘nRecordLength’ from ‘uint64_t’ {aka ‘long long unsigned int’} to ‘std::size_t’ {aka ‘unsigned int’} inside { } [-Werror=narrowing]
         annex.read(AsWritableBytes(Span{&vRecordValue, nRecordLength}));
```

ariard commented:
Yeah pushed something new. I'll add decent functional test coverage at least.

ajtowns commented on src/consensus/params.h:
```diff
@@ -88,6 +90,8 @@ struct Params {
     /** Don't warn about unknown BIP 9 activations below this height.
      * This prevents us from warning about the CSV and segwit activations. */
+    /** Block height at which annex (BIPXXX) becomes active */
+    int AnnexHeight = std::numeric_limits<int>::max();
```
`FooHeight` is for deployments that are active on mainnet; they're not needed for `DeploymentPos` (the height is determined dynamically by signalling)

ariard replied:
Corrected.

ajtowns replied:
Still plenty of `AnnexHeight` references; maybe run "make" before "git push" ? :wink:

ariard replied:
Yeah lol, run make this time and grep for `AnnexHeight` references.

ariard commented:
Another use-case, contract-level timelock.

```
15:14 < instagibbs> one edge-case I am thinking about: you update a channel with your offers/fulfillments, then commit to the update. The receiving peer then witholds their partial signature, disconnects. You 
                    eventually go to chain unilaterally. Counterparty waits shared_delay - 1 blocks, then gets last update mined. shared_delay has effectively doubled, completely under the control of 
                    counterparty
15:18 < instagibbs> one-shot architectures fix this by being 2-party and one shot, or we'd have to "pass on" the already passed blocks to decrease later shared_delay timeouts, which I know there was brief 
                    discussion about
16:45 < instagibbs> ariard, using annex to commit to coin age being (less than) or equal to the field would then allow this in principle, I think. handwave covenants
17:33 < instagibbs> actually, this is awfully difficult to do with the keeping valid transactions valid, or avoiding messing with utxo indexing. Sigh.
Day changed to 20 Oct 2022
01:42 < ariard> instagibbs: if you mean something like a shared_delay timelock where the endpoint isn't the previous parent but the "root" utxo of the contract, i believe you would need some new 
                "spent-but-marked" coin caches
01:42 < ariard> and not sure if would be compatible with node operators incentives, even if you pay an overhead fee cost. though marked utxo might expire after some "hard" limit of one month, idk
01:44 < ariard> iirc you might need such handwawy covenants for coinpool-style construction, to avoid overlap of relative-timelocked withdrawal tx, only 1st one should be delayed to allow contestation of 
                current update_tx by a higher-numbered one 
```

ariard commented:
Note to myself: another "could-be" use-case "amount wildcard" distribution in non-interactive channel (see https://github.com/ariard/bitcoin-contracting-primitives-wg/issues/19)

ajtowns commented:
Suggest targeting this against 24.0 branch. (Apparently you can do that via the [Edit] button!)

bigspider commented:
Great to see this, as I think this solves in an elegant way the one missing bit in the specs of the [MATT proposal](https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2022-November/021182.html): how to make sure that the witness elements that are part of the contract execution are not malleated? For example, [this game](https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2022-November/021205.html) requires parties to commit to a number, which should therefore be signed or miners could just change it. Those elements can be thought of as "arguments" during the contract execution.

With a PUSH_ANNEX_RECORD opcode, this would allow to either:
- put the otherwise-malleable arguments in the annex instead of the witness, so that they are already signed when spending the covenant UTXO, or
- put a single hash in the annex, and then use it as a commitment for all the witness arguments that need to be made non-malleable (using OP_CAT or OP_SHA256CAT − which is already part of the proposal − in order to allow a single hash to commit to multiple elements).

Either way, this is clearly more efficient than the alternative solution of adding OP_CHECKSIGFROMSTACK, which would require one more signature check.

naumenkogs commented on src/script/interpreter.cpp:
```diff
@@ -2061,6 +2062,70 @@ uint256 ComputeTaprootMerkleRoot(Span<const unsigned char> control, const uint25
+         * 40456331200 records with 4 bytes.
+         * See `VarIntMode` comment in serialize.h for more.
+         */
+        uint64_t nRecordHigherMask = 0;
```
1. What does "higher-mask type" mean?
2. Why using the ugly variable naming notation? :)

instagibbs replied:
would like some text on the masking myself

ariard replied:
Completely updated the code here, there should be new text. Dropped "higher-mask type" reference. On the ugly variable naming notation, sounds to me the one in vigor for a bunch of `interpreter.cpp`, see `CheckSequence()` though feel free to suggest something else.

naumenkogs commented:
Great to have this as a platform for experiments, looking forward to prototypes on top of it

instagibbs commented on src/script/interpreter.cpp:
```diff
@@ -2061,6 +2062,70 @@ uint256 ComputeTaprootMerkleRoot(Span<const unsigned char> control, const uint25
     return k;
 }
 
+bool VerifyAnnex(const std::vector<unsigned char>& annex_stack, ScriptExecutionData& execdata)
```
`annex_vec`?

ariard replied:
Yep took it normally.

instagibbs commented on src/script/interpreter.cpp:
```diff
@@ -2061,6 +2062,70 @@ uint256 ComputeTaprootMerkleRoot(Span<const unsigned char> control, const uint25
+     * `VarInt` OR 0 bytes: length
+     * `length`-bytes: value
+     */
+    while (!annex_stack.empty()) {
```
I don't know what this loop is doing over the const reference to the annex_stack and I'm afraid to ask

ariard replied:
Friendly reminder I should add a dummy example for a annex entry and functional test :p

instagibbs commented on src/script/interpreter.cpp:
```diff
@@ -2061,6 +2062,70 @@ uint256 ComputeTaprootMerkleRoot(Span<const unsigned char> control, const uint25
+        uint8_t nRecordLowerMask;
+        annex.read(AsWritableBytes(Span{&nRecordLowerMask, 1}));
+
+        const uint64_t nRecordType = nRecordHigherMask * 64 + (nRecordLowerMask & 0x3F);
```
Please make constants for masks with explanations what they're for

ariard replied:
More comments, see the pending question in the BIP.

ariard commented:
Updated at 4ccd693. Thanks for the reviews!

Still I'll think to add a dummy annex entry and more functional tests. See the BIP for open questions.
 
>    put the otherwise-malleable arguments in the annex instead of the witness, so that they are already signed when spending the covenant UTXO, or
>    put a single hash in the annex, and then use it as a commitment for all the witness arguments that need to be made non-malleable (using OP_CAT or OP_SHA256CAT − which is already part of the proposal − in order to allow a single hash to commit to multiple elements).

If the "arguments" are bitcoin objects like curve points, amounts or scriptpubkeys, introducing in the annex might allow more flexibility for other use-cases. I don't know if the second approach is not more verbose (though you only need few "hash" annex entries).

naumenkogs commented on src/script/interpreter.cpp:
```diff
@@ -2061,6 +2062,52 @@ uint256 ComputeTaprootMerkleRoot(Span<const unsigned char> control, const uint25
+            default:
+                return true;
+        }
+	nLastRecordType = nRecordType;
```
This variable could be easily removed altogether (it is pretty redundant), unless you think it improves readability.

ariard replied:
changed on the latest push.

naumenkogs commented on src/script/interpreter.cpp:
```diff
@@ -2061,6 +2062,52 @@ uint256 ComputeTaprootMerkleRoot(Span<const unsigned char> control, const uint25
+	if (nRecordLength == 0x7F) {
+		uint64_t nBuffRecordLength = 0;
+		annex >> VARINT(nBuffRecordLength);
+		nRecordLength += nBuffRecordLength;
```
better reads if `nRecordLength = 0x7F + nBuffRecordLength`

ariard replied:
note this is a bitmask above not an assignment

naumenkogs commented:
There's a bunch of code improvements you can make: snake case, `uint xyz{0}` instead of `uint xyz = 0`, marking `const`... I assume you're aware and I won't be mentioning anymore. The ones above I think just help to read the code. Ah, adding more comments would also help.

>Still I'll think to add a dummy annex entry and more functional tests.

This would be great!

>See the BIP for open questions.

I don't see any open questions in the [BIP](https://github.com/ariard/bips/blob/2022-07-bip-annex/bip-annex.mediawiki).

>If the "arguments" are bitcoin objects like curve points, amounts or scriptpubkeys, introducing in the annex might allow more flexibility for other use-cases.

Can you elaborate on this? The "other use-cases" I mean. Other from what?

instagibbs commented:
a basic test case would be great for me to trace through for better understanding fwiw

ariard commented:
> I don't see any open questions in the [BIP](https://github.com/ariard/bips/blob/2022-07-bip-annex/bip-annex.mediawiki).

The BIP PR opens against the repository: https://github.com/bitcoin/bips/pull/1381

> Can you elaborate on this? The "other use-cases" I mean. Other from what?

One use-case you can think of is delegation with "blessed keys" committed as annex record. Alice signs a annex with Bob's key, allowing Bob to spend the output at anytime without Alice further interactivity.

> a basic test case would be great for me to trace through for better understanding fwiw

Working on the "block-feerate lock-point" mentioned above: https://github.com/ariard/bitcoin/tree/2023-01-feerate-point, as  a solid example of what the annex can be useful for :)

ariard committed:
Add VerifyAnnex

ariard committed:
Add annex record enum and AnnexValidationResult

ariard committed:
Add IsAnnexStandard

ariard committed:
Add SCRIPT_VERIFY_ANNEX deployment code

ariard commented:
Updated at ba8569c77:
- introduce a `AnnexValidationResult` to enforce policy-only rejection of “unknown records” annex
- introduce a `ANNEX_RECORD_POLICY_RESERVED` to allow annex carrier for unstructured data preserving future evolvability
- rebased on top 24.0 and #22
